### ISSUE TITLE
Grails assetCompile fails with "Too many open files in system" for large numbers of assets

### ISSUE THREAD

**Issue by StevenACoffman** [bertramdev/asset-pipeline#78]
I have been migrating a large legacy grails codebase to grails 3.x. There are over 3500 assets (note: I inherited this, not my fault) and when I run `gradle assetCompile` it fails with "Too many open files in system". It appears that something is not closing the file streams after a file has been written.

I have [a gist with the stacktrace](https://gist.github.com/StevenACoffman/9bfd1a1b1ac72bee4b08) if that is helpful. I am using gradle 2.11, with asset-pipeline-gradle 2.6.9, grails 3.0.14 on a Mac.

When I run:

```bash
 sysctl kern.maxfiles; sysctl kern.maxfilesperproc; ulimit -a
```

I get:
```bash
kern.maxfiles: 12288
kern.maxfilesperproc: 10240
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
file size               (blocks, -f) unlimited
max locked memory       (kbytes, -l) unlimited
max memory size         (kbytes, -m) unlimited
open files                      (-n) 10240
pipe size            (512 bytes, -p) 1
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 709
virtual memory          (kbytes, -v) unlimited
```

**Comment by davydotcom** [comment]
Well that's very odd I'll hunt

**Comment by StevenACoffman** [comment]
Any ideas? Kinda hurting right now.

**Comment by davydotcom** [comment]
i found the cause , build forthcoming

**Comment by StevenACoffman** [comment]
You are awesome! Thank you so much for your work on this.

**Comment by davydotcom** [comment]
fixed in `asset-pipeline-gradle 2.7.0` and grails asset-pipeline `3.1.0`

**Comment by davydotcom** [comment]
just want to confirm this fix worked for you

**Comment by StevenACoffman** [comment]
Sometimes! Unfortunately, not very consistently, but I really appreciate your diligent effort. When it does fail, it seems to get farther before it fails. I set up a repository that can reproduce this problem at https://github.com/StevenACoffman/newt 

The newt repository uses grails 3.1.3, groovy 2.4.6, and gradle 2.9 (but I also tried gradle 2.11). My local machine is also using Java 1.8.0_66 on Mac OS X Yosemite. If I try either of the following:

```bash
rm -r -f build; grails dev war
rm -r -f build; gradle assetCompile
```
I still frequently get (the exact number of assets processed varies considerably) something like:

```bash
Writing File 3190 of 3822 - icon/fugue/arrow-branch-270-left
:assetCompile FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':assetCompile'.
> java.io.IOException: Too many open files in system

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 18.264 secs
```

My gradle.properties file looks like this:

```bash
org.gradle.daemon=true
org.gradle.jvmargs =-Xmx4048m -Xms512m
```

I've tried omitting or tweaking the jvmargs line and it doesn't seem to help. I've also tried setting the daemon to false, and that didn't seem to help either.

With your improvement, I can now work around the problem by just trying repeatedly, where it never succeeded before. I wanted to follow up with a reproducible case, as it might be a sign of an issue that affects those with even larger numbers of assets (probably rare), or might get worse with future changes.

What is exceptionally odd is that other developers on my team don't appear to have this problem on either linux or windows, but they may have other environmental differences (jdk, gradle settings, etc.).

**Comment by davydotcom** [comment]
Ah gotya well the good news is you confirmed I did close a leak but sounds like there may be another straggler in there. It's possible your particular machine already is using a lot of file descriptors that is making this more prominent as well. I have my system set to unlimited now but with Yosemite they lowered the default limit to like 1024 open files unless you did an override.

**Comment by davydotcom** [comment]
Just checking this sample project only has 500 assets

**Comment by StevenACoffman** [comment]
Argh. Sorry. I had .gitignore'd a bit too aggressively and so the icons all were excluded from my commit.

**Comment by davydotcom** [comment]
Published 2.7.2 and 3.1.1(grails3) should have further reduced that issue, I noticed a significant bump in compile time performance on your newt project as well but was unable to reproduce the open file handles error

**Comment by StevenACoffman** [comment]
So the Yosemite file descriptors was to blame, and fixing that actually helped a couple other problems on my system, so thanks! Like you said, the new version screams, and I can't get it to fail anymore. Thanks again!
